### ISSUE TITLE
AttributeError: Either 'Sale' object has no attribute 'balancedue' ... caused by ArTran with single space paymenttypeid

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2080]
_Imported from trac ticket 2080.
Created by **kbower**
Opened in trac: Jan 22, 2013, 8:34:55 AM
Last modified in trac: Dec 15, 2016, 4:42:44 PM_
***
When attempting to bring up a customer history or an order, this error is encountered: `AttributeError: Either 'Sale' object has no attribute 'balancedue' or evaluating 'balancedue' caused an AttributeError`

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 22, 2013, 8:36:27 AM:_

Wayne,
    The root cause here is that archived AR transactions exist with a single space ' ' for a MOP_CD. 

```
  1* select CUST_CD, TRN_TP_CD, '['||MOP_CD||']' from AR_TRN_ARC 
where CUST_CD = 'CHIAM16593'
SQL> /

CUST_CD    TRN '['||
---------- --- -----
CHIAM16593 DEP [DC]
CHIAM16593 SAL []
CHIAM16593 DEP [AMX]
CHIAM16593 SAL []
CHIAM16593 DEP [MC]
CHIAM16593 SAL []
CHIAM16593 DEP [BA]
CHIAM16593 SAL [ ]
CHIAM16593 SAL [ ]
CHIAM16593 DEP [BA]
CHIAM16593 DTD []
CHIAM16593 DTC []

12 rows selected.

SQL> select count(*) from AR_TRN_ARC where MOP_CD = ' ';

  COUNT(*)
----------
    103824

SQL> select count(*) from AR_TRN where MOP_CD = ' ';

  COUNT(*)
----------
        39
```

We saw this at El Dorado as well.  Do you know the cause of this?

Clearly ' ' is not a valid method of payment:
```
SQL> select  '['||MOP_CD||']' from mop;

'['||
-----
[CS]
[CK]
[AMX]
[MC]
[DC]
[BA]
[FI]
[COD]
[OA]
[PC]
[RC]
[WCI]
[NC]
[DB]
[BF]
[GC]
[JPC]
[TW]
[GE]

19 rows selected.
```

But Loft expects a foreign key relationship to exist there with data integrity.  Any ideas? 

My largest concern is that there are current records of this type in in AR_TRN.  How did these get there?

Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 22, 2013, 9:00:40 AM:_

This happens for all 6 current customers.  It appears to happen exclusively to 'SAL' transactions when it occurs.

Rather than add crud in Loft base code, consider AR_TRN insert trigger to fix this on GERS.  I can't imagine the problem occurs in AR_TRN_ARC, which should just be copied data from AR_TRN.

Hopefully this isn't some cryptic intentional data tracking by GERS that has meaning.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 22, 2013, 9:55:10 AM:_

apparently GERS fixed bug, all such records are many years old.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 22, 2013, 9:55:44 AM:_

(Changeset [[3573]](../commit/0c05933060c47c5bc44928818e819029ce12ee00))
* update artrans and artransarchive's paymenttypeid to null where it is ' ' (fixes #2080)

**Comment by kentbower** [comment]
_**kbower** trac status change on Jan 22, 2013, 9:55:44 AM:_ **closed**
